### ISSUE TITLE
Unable to Load Models from Azure Storage

### ISSUE THREAD

**Issue by debugaftercoffee** [triton-inference-server/server#5533]
**Description**
I am running the latest TritonServer Docker container on a Linux VM and I'm following these [instructions]( https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_repository.md#azure-storage) to use Azure Storage as the model repository. The Triton Server never finishes loading and when I enable debug logging, the server is in an infinite loop trying to list the model artifacts. It looks like it is trying the model artifacts as a directory instead of a file.

**Triton Information**
I'm using the official Docker container: nvcr.io/nvidia/tritonserver:23.02-py3

**To Reproduce**
I have a Onnx model loaded in Azure Storage following the model repository directory structure. In the storage container _triton_, I have a top-level directory called _models_, and in the _models_ directory a have a single model called _text_classifier_. If I use azcopy to list the container contents this is what I see.
 
***Contents of Azure Blob Storage***
```
$ azcopy list https://mystorage.blob.core.windows.net/triton
INFO: models/text_classifier/config.pbtxt;  Content Length: 355.00 B
INFO: models/text_classifier/1/model.onnx;  Content Length: 255.54 MiB
```
 
 
Below is the Docker command I am using to startup the container; I am setting the “AZURE_STORAGE_ACCOUNT” and “AZURE_STORAGE_KEY” environment variables. I have also tried the preview feature “TRITON_CLOUD_CREDENTIAL_PATH” environment variable which leads to the same issue. The behavior is the container starts but does not load any models. When I enable debug logging, I can see Triton is in an infinite loop trying to list the contents of the model.onnx file as if it was a directory. I have verified the account key I’m using is valid, if I use the wrong account key Triton stops immediately with an authorization error. The issue look similar to this [issue](https://github.com/triton-inference-server/server/issues/5501). I have tried older versions of the Triton Server container as far back as a year ago and the issue is reproducible. If I use the [local file system](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_repository.md#local-file-system) model repository, my model loads and inferencing works as expected. 
 
***Docker Command***
```
$ docker run --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864 --gpus=1 --rm --net=host --name tritonserver \
> -e AZURE_STORAGE_ACCOUNT=mystorage \
> -e AZURE_STORAGE_KEY=<MY_ACCOUNT_KEY_REDACTED> \
> nvcr.io/nvidia/tritonserver:23.02-py3 tritonserver --model-repository=as://mystorage/triton/models --log-verbose=3 --log-info=1 --log-warning=1 --log-error=1
 
=============================
== Triton Inference Server ==
=============================
 
NVIDIA Release 23.02 (build 53616260)
Triton Server Version 2.31.0
 
Copyright (c) 2018-2023, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
 
Various files include modifications (c) NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
 
This container image and its contents are governed by the NVIDIA Deep Learning Container License.
By pulling and using the container, you accept the terms and conditions of this license:
https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license
 
I0316 14:22:55.926700 1 pinned_memory_manager.cc:240] Pinned memory pool is created at '0x7ff610000000' with size 268435456
I0316 14:22:55.933096 1 cuda_memory_manager.cc:105] CUDA memory pool is created on device 0 with size 67108864
I0316 14:22:55.942137 1 filesystem.cc:2276] TRITON_CLOUD_CREDENTIAL_PATH environment variable is not set, reading from environment variables
I0316 14:22:55.942165 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models
I0316 14:22:56.099343 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models
I0316 14:22:56.138704 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier
I0316 14:22:56.163525 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier
I0316 14:22:56.187687 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier
I0316 14:22:56.213874 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1
I0316 14:22:56.242835 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1
I0316 14:22:56.270226 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1
I0316 14:22:56.294956 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx
I0316 14:22:56.328797 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx
I0316 14:22:56.354268 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx
I0316 14:22:56.391336 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.416799 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.438017 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.464517 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.490563 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.514786 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.558477 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.568352 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.572585 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.580891 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.590927 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
I0316 14:22:56.595072 1 filesystem.cc:2354] Using credential    for path  as://mystorage/triton/models/text_classifier/1/model.onnx/
```

**Expected behavior**
The Onnx model in Blob Storage should be loaded by the Triton Server.

**Comment by jbkyang-nvi** [comment]
@kthui can you take a look?

**Comment by kthui** [comment]
Hi @debugaftercoffee, can you try setting [--model-control-mode=explicit](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_management.md#model-control-mode-explicit) when starting the server, and then load a small model (small in file size) using the [load API](https://github.com/triton-inference-server/client/blob/main/src/python/library/tritonclient/http/__init__.py#L641) from the client, and see if the hang is still reproducible? It is because the model(s) might just takes a long time to be transmitted over the internet.

**Comment by kthui** [comment]
Hi @debugaftercoffee, thanks for the update. I was able to identify the loop that caused the infinite access to `as://mystorage/triton/model_repo/debug/1/model.py/`. However, I was not able to replicate the issue with our own Azure credential. Would you be able to give the following a try?
1. Can you try `listing` and `downloading` the `as://mystorage/triton/model_repo/debug/1/model.py` file using the [Azure command line](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-cli#list-the-blobs-in-a-container), with the credential supplied to Triton? If this is not working, then it indicates some permission is missing from the Azure credential.
2. If the above command line worked, can you try starting Triton with the root credential from Azure (the credential that has full permission into the container)? If the root credential worked, then it indicates some lack of permission, but not reproducible by the command line argument. If it does not work, please let us know and I can file a ticket for us to investigate deeper into the Azure client.

**Comment by debugaftercoffee** [comment]
@kthui I was able to list and download the model file using the `az storage blob` commands. I set `--auth-mode key` and `--acount-key` with the `az storage blob` commands to specify the Storage Account Key that Triton is using. The key I am using is the Storage Account Access key which has full permissions on the container.

**Comment by kthui** [comment]
Thanks for the update. I have filed a ticket for us to investigate further. 

Could you give [r23.03](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tritonserver/tags) a try? In case this is a version specific issue.

**Comment by debugaftercoffee** [comment]
@kthui I tried [r23.03](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tritonserver/tags) and I get the same issue. I think I narrowed down a Storage Account setting that is causing the problem. The Storage Account I am using has [ADLSGen2](https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction) enabled on it. I created a new Storage Account and set `Enable hierarchical namespace=False` and Triton is able to load models from Azure Storage Account. If you create a Storage Account and set `Enable hierarchical namespace=True`, you will see the problem.
